File: src/cpp_module/gif.cpp
// Python の pylib.gif に対応する GIF 書き出し実装です。

#include "cpp_module/gif.h"

#include <cstdint>
#include <fstream>
#include <stdexcept>
#include <vector>

namespace pytra::cpp_module::gif {
namespace {

void append_u16_le(std::vector<std::uint8_t>& out, std::uint16_t v) {
    out.push_back(static_cast<std::uint8_t>(v & 0xFF));
    out.push_back(static_cast<std::uint8_t>((v >> 8) & 0xFF));
}

void append_bytes(std::vector<std::uint8_t>& out, const std::uint8_t* data, std::size_t len) {
    out.insert(out.end(), data, data + len);
}

void emit_code(
    std::vector<std::uint8_t>& out,
    std::uint32_t& bit_buffer,
    int& bit_count,
    int code_size,
    int code
) {
    bit_buffer |= (static_cast<std::uint32_t>(code) << bit_count);
    bit_count += code_size;
    while (bit_count >= 8) {
        out.push_back(static_cast<std::uint8_t>(bit_buffer & 0xFF));
        bit_buffer >>= 8;
        bit_count -= 8;
    }
}

std::vector<std::uint8_t> lzw_encode(const std::vector<std::uint8_t>& data, int min_code_size) {
    if (data.empty()) {
        return {};
    }

    const int clear_code = 1 << min_code_size;
    const int end_code = clear_code + 1;
    int code_size = min_code_size + 1;

    std::vector<std::uint8_t> out;
    // 現実装は各ピクセルごとに clear code を挟むため、出力量は入力より大きくなる。
    // 十分な容量を先に確保し、LTO 環境での再確保由来警告とオーバーヘッドを避ける。
    out.reserve(data.size() * 3 + 16);

    std::uint32_t bit_buffer = 0;
    int bit_count = 0;

    emit_code(out, bit_buffer, bit_count, code_size, clear_code);
    code_size = min_code_size + 1;

    for (std::uint8_t c : data) {
        emit_code(out, bit_buffer, bit_count, code_size, static_cast<int>(c));
        emit_code(out, bit_buffer, bit_count, code_size, clear_code);
        code_size = min_code_size + 1;
    }
    emit_code(out, bit_buffer, bit_count, code_size, end_code);

    if (bit_count > 0) {
        out.push_back(static_cast<std::uint8_t>(bit_buffer & 0xFF));
    }

    return out;
}

}  // namespace

std::vector<std::uint8_t> grayscale_palette() {
    std::vector<std::uint8_t> p;
    p.reserve(256 * 3);
    for (int i = 0; i < 256; ++i) {
        const auto v = static_cast<std::uint8_t>(i);
        p.push_back(v);
        p.push_back(v);
        p.push_back(v);
    }
    return p;
}

void save_gif(
    const std::string& path,
    int width,
    int height,
    const std::vector<std::vector<std::uint8_t>>& frames,
    const std::vector<std::uint8_t>& palette,
    int delay_cs,
    int loop
) {
    if (width <= 0 || height <= 0) {
        throw std::runtime_error("gif: width/height must be positive");
    }
    if (palette.size() != 256 * 3) {
        throw std::runtime_error("gif: palette must be 768 bytes");
    }

    const std::size_t frame_bytes = static_cast<std::size_t>(width) * static_cast<std::size_t>(height);
    for (const auto& fr : frames) {
        if (fr.size() != frame_bytes) {
            throw std::runtime_error("gif: frame size mismatch");
        }
    }

    std::vector<std::uint8_t> out;
    out.reserve(static_cast<std::size_t>(1024) + frames.size() * frame_bytes / 2);

    append_bytes(out, reinterpret_cast<const std::uint8_t*>("GIF89a"), 6);
    append_u16_le(out, static_cast<std::uint16_t>(width));
    append_u16_le(out, static_cast<std::uint16_t>(height));
    out.push_back(static_cast<std::uint8_t>(0xF7));
    out.push_back(static_cast<std::uint8_t>(0));
    out.push_back(static_cast<std::uint8_t>(0));
    append_bytes(out, palette.data(), palette.size());

    out.push_back(static_cast<std::uint8_t>(0x21));
    out.push_back(static_cast<std::uint8_t>(0xFF));
    out.push_back(static_cast<std::uint8_t>(0x0B));
    append_bytes(out, reinterpret_cast<const std::uint8_t*>("NETSCAPE2.0"), 11);
    out.push_back(static_cast<std::uint8_t>(0x03));
    out.push_back(static_cast<std::uint8_t>(0x01));
    append_u16_le(out, static_cast<std::uint16_t>(loop));
    out.push_back(static_cast<std::uint8_t>(0x00));

    for (const auto& fr : frames) {
        out.push_back(static_cast<std::uint8_t>(0x21));
        out.push_back(static_cast<std::uint8_t>(0xF9));
        out.push_back(static_cast<std::uint8_t>(0x04));
        out.push_back(static_cast<std::uint8_t>(0x00));
        append_u16_le(out, static_cast<std::uint16_t>(delay_cs));
        out.push_back(static_cast<std::uint8_t>(0x00));
        out.push_back(static_cast<std::uint8_t>(0x00));

        out.push_back(static_cast<std::uint8_t>(0x2C));
        append_u16_le(out, static_cast<std::uint16_t>(0));
        append_u16_le(out, static_cast<std::uint16_t>(0));
        append_u16_le(out, static_cast<std::uint16_t>(width));
        append_u16_le(out, static_cast<std::uint16_t>(height));
        out.push_back(static_cast<std::uint8_t>(0x00));

        out.push_back(static_cast<std::uint8_t>(0x08));
        const std::vector<std::uint8_t> compressed = lzw_encode(fr, 8);
        std::size_t pos = 0;
        while (pos < compressed.size()) {
            const std::size_t len = (compressed.size() - pos > 255) ? 255 : (compressed.size() - pos);
            out.push_back(static_cast<std::uint8_t>(len));
            append_bytes(out, compressed.data() + pos, len);
            pos += len;
        }
        out.push_back(static_cast<std::uint8_t>(0x00));
    }

    out.push_back(static_cast<std::uint8_t>(0x3B));

    std::ofstream ofs(path, std::ios::binary);
    if (!ofs) {
        throw std::runtime_error("gif: failed to open output file");
    }
    ofs.write(reinterpret_cast<const char*>(out.data()), static_cast<std::streamsize>(out.size()));
}

}  // namespace pytra::cpp_module::gif

File: src/pytra/utils/gif.py
"""アニメーションGIFを書き出すための最小ヘルパー。"""

from __future__ import annotations

from pytra.std import abi


def _gif_append_list(dst: list[int], src: list[int]) -> None:
    i = 0
    n = len(src)
    while i < n:
        dst.append(src[i])
        i += 1


def _gif_u16le(v: int) -> list[int]:
    return [v & 0xFF, (v >> 8) & 0xFF]


def _lzw_encode(data: bytes, min_code_size: int = 8) -> bytes:
    if len(data) == 0:
        return bytes([])

    clear_code = 1 << min_code_size
    end_code = clear_code + 1
    code_size = min_code_size + 1

    out: list[int] = []
    bit_buffer = 0
    bit_count = 0

    bit_buffer |= clear_code << bit_count
    bit_count += code_size
    while bit_count >= 8:
        out.append(bit_buffer & 0xFF)
        bit_buffer = bit_buffer >> 8
        bit_count -= 8
    code_size = min_code_size + 1

    for v in data:
        bit_buffer |= v << bit_count
        bit_count += code_size
        while bit_count >= 8:
            out.append(bit_buffer & 0xFF)
            bit_buffer = bit_buffer >> 8
            bit_count -= 8

        bit_buffer |= clear_code << bit_count
        bit_count += code_size
        while bit_count >= 8:
            out.append(bit_buffer & 0xFF)
            bit_buffer = bit_buffer >> 8
            bit_count -= 8

        code_size = min_code_size + 1

    bit_buffer |= end_code << bit_count
    bit_count += code_size
    while bit_count >= 8:
        out.append(bit_buffer & 0xFF)
        bit_buffer = bit_buffer >> 8
        bit_count -= 8

    if bit_count > 0:
        out.append(bit_buffer & 0xFF)

    return bytes(out)


def grayscale_palette() -> bytes:
    p: list[int] = []
    i = 0
    while i < 256:
        p.append(i)
        p.append(i)
        p.append(i)
        i += 1
    return bytes(p)


@abi(args={"frames": "value"})
def save_gif(
    path: str,
    width: int,
    height: int,
    frames: list[bytes],
    palette: bytes,
    delay_cs: int = 4,
    loop: int = 0,
) -> None:
    if len(palette) != 256 * 3:
        raise ValueError("palette must be 256*3 bytes")

    frame_lists: list[list[int]] = []
    for fr in frames:
        fr_list: list[int] = []
        for v in fr:
            fr_list.append(int(v))
        if len(fr_list) != width * height:
            raise ValueError("frame size mismatch")
        frame_lists.append(fr_list)

    palette_list: list[int] = []
    for v in palette:
        palette_list.append(int(v))

    out: list[int] = []
    _gif_append_list(out, [71, 73, 70, 56, 57, 97])  # GIF89a
    _gif_append_list(out, _gif_u16le(width))
    _gif_append_list(out, _gif_u16le(height))
    out.append(0xF7)
    out.append(0)
    out.append(0)
    _gif_append_list(out, palette_list)

    _gif_append_list(out, [0x21, 0xFF, 0x0B, 78, 69, 84, 83, 67, 65, 80, 69, 50, 46, 48, 0x03, 0x01])
    _gif_append_list(out, _gif_u16le(loop))
    out.append(0)

    for fr_list in frame_lists:
        _gif_append_list(out, [0x21, 0xF9, 0x04, 0x00])
        _gif_append_list(out, _gif_u16le(delay_cs))
        _gif_append_list(out, [0x00, 0x00])

        out.append(0x2C)
        _gif_append_list(out, _gif_u16le(0))
        _gif_append_list(out, _gif_u16le(0))
        _gif_append_list(out, _gif_u16le(width))
        _gif_append_list(out, _gif_u16le(height))
        out.append(0)
        out.append(8)
        compressed = _lzw_encode(bytes(fr_list), 8)
        pos = 0
        while pos < len(compressed):
            remain = len(compressed) - pos
            chunk_len = 255 if remain > 255 else remain
            out.append(chunk_len)
            i = 0
            while i < chunk_len:
                out.append(compressed[pos + i])
                i += 1
            pos += chunk_len
        out.append(0)

    out.append(0x3B)

    with open(path, "wb") as f:
        f.write(bytes(out))
